pkg/cover/file.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package cover
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"cloud.google.com/go/civil"
+	"github.com/google/syzkaller/pkg/covermerger"
+)
+
+type lineRender func(string, bool, int, int) string
+
+func RendFileCoverage(c context.Context, ns, repo, commit, filePath string,
+	fromDate, toDate civil.Date, render lineRender) (string, error) {
+	fileContent, err := covermerger.GetFileVersion(filePath, repo, commit)
+	if err != nil {
+		return "", fmt.Errorf("failed to GetFileVersion for file %s, commit %s from repo %s: %w",
+			filePath, commit, repo, err)
+	}
+	config := &covermerger.Config{
+		Jobs: 1,
+		Base: covermerger.RepoBranchCommit{
+			Repo:   repo,
+			Commit: commit,
+		},
+		FileVersProvider: covermerger.MakeWebGit(),
+	}
+
+	dbReader := covermerger.MakeBQCSVReader()
+	if err := dbReader.InitNsRecords(c,
```

Comment:
@dvyukov let's make the nil deref discussion threaded.
It fails under syz-env and doesn't fail w/o it.
Could you please `gcloud auth login` and `gcloud auth application-default login` on your machine?
Then `go run ./tools/syz-cover -for-file virt/kvm/binary_stats.c -from 2024-08-01 -to 2024-08-10`. The commit is not needed now.

Reply:
#5180 to fix the problem. We'll see better error description then.

Reply:
It worked for me after the gcloud commands. Perhaps it's worth to detect and suggest in a failure message (instead of nil deref).

The report looks strange. I've tried:
```
$ go run ./tools/syz-cover -for-file fs/open.c -from 2024-08-01 -to 2024-08-10
```

And I see lots of coverage inside of comments and between functions:
```
     0    708   }
          709   return error;
          710 }
     0    711 
     0    712 SYSCALL_DEFINE4(fchmodat2, int, dfd, const char __user *, filename,
          713           umode_t, mode, unsigned int, flags)
  2754    714 {
     0    715   return do_fchmodat(dfd, filename, mode, flags);
...
          729 /*
          730  * Check whether @kuid is valid and if so generate and set vfsuid_t in
          731  * ia_vfsuid.
          732  *
     0    733  * Return: true if @kuid is valid, false if not.
          734  */
  1599    735 static inline bool setattr_vfsuid(struct iattr *attr, kuid_t kuid)
          736 {
    36    737   if (!uid_valid(kuid))
...
          744 /*
          745  * Check whether @kgid is valid and if so generate and set vfsgid_t in
          746  * ia_vfsgid.
          747  *
     0    748  * Return: true if @kgid is valid, false if not.
          749  */
  1755    750 static inline bool setattr_vfsgid(struct iattr *attr, kgid_t kgid)
...
          800                   goto retry_deleg;
          801   }
          802   return error;
   457    803 }
    11    804 
          805 int do_fchownat(int dfd, const char __user *filename, uid_t user, gid_t group,
          806           int flag)
  6769    807 {
  3314    808   struct path path;
          809   int error = -EINVAL;
          810   int lookup_flags;
    11    811 
    91    812   if ((flag & ~(AT_SYMLINK_NOFOLLOW | AT_EMPTY_PATH)) != 0)
     0    813           goto out;
```

Is there a bug somewhere? Or it's working as intended?

Reply:
Which makes me wonder if we aggregate correctly.
Total coverage for this file based on the report is 680/(680+122)=~0.85%.
syzbot aggregated numbers also show ~85% for this file recently.
However, if I look at recent individual coverage reports from common instances I see 61%, 58%, 57%. Aggregated coverage should be larger, but I am not sure it can be that much larger for this file (not arch-specific or anything).
Is it possible that we merge mismatching coverage from other instances (on empty lines/comments/etc), but don't add mismatching uncovered lines, and as the result total coverage is falsely large.

Reply:
For some files/commits it works correct. For some it doesn't.
My observations are the same as yours. I intentionally used virt/kvm/binary_stats.c because it looks Ok.
For virt/kvm/kvm_main.c it mostly works correct but there are some crazy points at the end of the file.
This tool will be used to detect guilty file commit pairs and improve fix/improve it.

Reply:
Ack.
I would assume that the files for which it appears to work just have not changed within the range, so all lines match regardless of revisions.

Reply:
Will be fixed by #5188.